src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java

```diff
@@ -140,99 +133,101 @@ public void checkout(
-            // shared libraries need to set the workspace name every time
-            // because they use a common SCM object that isn't initialized
-            // in a sane way
-            if (isSharedLibrary(workspace)) {
-                workspaceInfo.setWorkspaceName("shl-" + workspace.getRemote().hashCode());
-                workspaceInfo.setDirectory("");
-            }
 
-            String plasticWorkspaceName = resolveWorkspaceNameParameters(workspace, workspaceInfo, run);
             FilePath plasticWorkspacePath = resolveWorkspacePath(workspace, workspaceInfo);
+            String resolvedSelector = SelectorParametersResolver.resolve(workspaceInfo.getSelector(), parameterValues);
 
-            String resolvedSelector = SelectorParametersResolver.resolve(
-                    workspaceInfo.getSelector(), parameterValues);
-            result.addAll(
-                setupWorkspace(
-                    run,
-                    launcher,
+            PlasticTool tool = new PlasticTool(getDescriptor().getCmExecutable(), launcher, listener, plasticWorkspacePath);
+
+            Workspace plasticWorkspace = setupWorkspace(
+                    tool,
                     listener,
-                    plasticWorkspaceName,
                     plasticWorkspacePath,
                     resolvedSelector,
-                    workspaceInfo.getUseUpdate())
-            );
+                    workspaceInfo.getUseUpdate());
+
+            ChangeSetID csetId = determineCurrentChangeset(tool, listener, plasticWorkspacePath);
+
+            ChangeSet cset = retrieveChangesetDetails(tool, listener, csetId.getId());
+            cset.setRepoName(csetId.getRepository());
+            cset.setRepoServer(csetId.getHost());
+
+            ChangeSet previousCset = retrievePreviouslyBuildChangeset(run, cset);
+            if (previousCset == null) {
+                changeLogItems.add(cset);
+            } else {
+                List<ChangeSet> changeSetItems = retrieveMultipleChangesetDetails(tool, listener,
+                        previousCset.getId(), cset.getId());
+                for (ChangeSet it : changeSetItems) {
+                    it.setRepoName(csetId.getRepository());
+                    it.setRepoServer(csetId.getHost());
+                }
+                changeLogItems.addAll(changeSetItems);
+            }
+
+            BuildData buildData = new BuildData(plasticWorkspace, cset);
+            List<BuildData> actions = run.getActions(BuildData.class);
+            if (!actions.isEmpty()) {
+                buildData.setIndex(actions.size()+1);
+            }
+            run.addAction(buildData);
         }
 
-        if (changelogFile != null)
-            writeChangeLog(listener, changelogFile, result);
+        if (changelogFile != null) {
+            writeChangeLog(listener, changelogFile, changeLogItems);
+        }
     }
 
     private static boolean isSharedLibrary(@Nonnull FilePath jenkinsPath) {
```

Comment:
I think this method is unused now, please remove it.

Reply:
I have left this method there as I wanted to use it in future modifications.

Reply:
Perfect, nothing else to remark then!